src/test/java/StringManipulationTest.java

```diff
@@ -119,31 +202,88 @@ public void testRestoreString1()
     }
 
     @Test
-    public void testRestoreString2()
+    //twenty-third test case
+    public void restoreString_stringAndIndicesLengthDoNotEqual_returnsIllegalArgumentException()
     {
-        fail("Not yet implemented");
+        manipulatedstring.setString("artistic");
+        int [] array = new int[]{1,0,2,3};
+        assertThrows(IllegalArgumentException.class, ()->
+                manipulatedstring.restoreString(array));
 
     }
 
     @Test
-    public void testRestoreString3()
+    //twenty-fourth test case
+    public void restoreString_indicesLessThan0_returnsIndexOutOfBoundsException()
     {
-        fail("Not yet implemented");
+        manipulatedstring.setString("art");
+        int [] array = new int[]{2,-1,0};
+        assertThrows(IndexOutOfBoundsException.class, ()->
+                manipulatedstring.restoreString(array));
+    }
 
+    @Test
+    //twenty-fifth test case
+    public void restoreString_indicesGreaterThanStringLength_returnsIndexOutOfBoundsException()
+    {
+        manipulatedstring.setString("art");
+        int [] array = new int[]{1,0,4};
+        assertThrows(IndexOutOfBoundsException.class, ()->
+                manipulatedstring.restoreString(array));
     }
 
     @Test
-    public void testRestoreString4()
+    //twenty-sixth test case
+    public void restoreString_nullString_returnNullPointerException()
     {
-        fail("Not yet implemented");
+        manipulatedstring.setString(null);
+        int [] array = new int[]{};
+        assertThrows(NullPointerException.class, ()->
+                manipulatedstring.restoreString(array));
+    }
+
+    @Test
+    //twenty-seventh test case
+    public void restoreString_emptyString_returnsEmptyString()
+    {
+        manipulatedstring.setString("");
+        int [] array;
+        array=new int[]{};
+        String restoreString = manipulatedstring.restoreString(array);
+        assertEquals(restoreString, "");
+    }
 
+    @Test
+    //twenty-eighth test case
+    public void restoreString_upperCaseLetters_returnsRestoredString()
+    {
+        manipulatedstring.setString("ART");
+        int [] array;
+        array=new int[]{1,0,2};
+        String restoreString = manipulatedstring.restoreString(array);
+        assertEquals(restoreString, "RAT");
     }
 
     @Test
-    public void testRestoreString5()
+    //twenty-ninth test case
+    public void restoreString_upperCaseAndLowerCaseLetters_returnsRestoredString()
```

Comment:
This test might seem be redundant after looking at the implementation, but I can understand why this was tested in the case that the implementation was different.